### PR TITLE
Extract collector DI and registration client to generic V3 project

### PR DESCRIPTION
**Pull request by joelverhagen**
This supports building other collectors that use dependency injection like Catalog2AzureSearch. Also, the registration client code is not Azure Search specific so a more generic place is better.

This is a PR to move code around that has no impact on behavior. It allows me to share code between other new collectors and Catalog2AzureSearch without pulling in all of the Azure Search dependencies.

### CODE REVIEW

File: src/NuGet.Services.AzureSearch/Catalog2AzureSearch/Catalog2AzureSearchCommand.cs [hunk 1]
```diff
@@ -11,6 +11,7 @@
 using Microsoft.WindowsAzure.Storage;
 using NuGet.Services.Metadata.Catalog;
 using NuGet.Services.Metadata.Catalog.Persistence;
+using NuGet.Services.V3;
```
Comment (loic-sharma) [inline src/NuGet.Services.AzureSearch/Catalog2AzureSearch/Catalog2AzureSearchCommand.cs]: This looks unused

File: src/NuGet.Services.AzureSearch/DependencyInjectionExtensions.cs [hunk 2]
```diff
@@ -216,31 +212,9 @@ private static void RegisterAuxiliaryDataStorageServices(ContainerBuilder contai
...
-                    AutomaticDecompression = DecompressionMethods.GZip | DecompressionMethods.Deflate
-                });
-
-            services
-                .AddTransient(p => (HttpMessageHandler)new TelemetryHandler(
-                    p.GetRequiredService<ITelemetryService>(),
-                    p.GetRequiredService<HttpClientHandler>()));
-
-            services.AddSingleton(p => new HttpClient(p.GetRequiredService<HttpMessageHandler>()));
+            services.AddV3();
```
Comment (loic-sharma) [inline src/NuGet.Services.AzureSearch/DependencyInjectionExtensions.cs]: 😍

File: src/NuGet.Services.V3/CommitCollectorUtility.cs [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+                if (!StringComparer.OrdinalIgnoreCase.Equals(max.PackageIdentity, entry.PackageIdentity))
+                {
+                    throw new InvalidOperationException("The entries compared should have the same package identity.");
+                }
+
+                if (entry.CommitTimeStamp > max.CommitTimeStamp)
+                {
+                    max = entry;
+                }
+                else if (entry.CommitTimeStamp == max.CommitTimeStamp)
```
Comment (loic-sharma) [inline src/NuGet.Services.V3/CommitCollectorUtility.cs]: Super nit, consider reducing nesting:

```csharp
if (entry.CommitTimeStamp == max.CommitTimeStamp)
{
    ...
}

if (entry.CommitTimeStamp > max.CommitTimeStamp)
{
    max = entry;
}
```

File: src/NuGet.Services.V3/CommitCollectorUtility.cs [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+        {
+            var workEnumerable = latestItems
+                .GroupBy(x => x.PackageIdentity.Id, StringComparer.OrdinalIgnoreCase)
+                .Select(x => new IdAndValue<IReadOnlyList<CatalogCommitItem>>(x.Key, x.ToList()));
+
+            var allWork = new ConcurrentBag<IdAndValue<IReadOnlyList<CatalogCommitItem>>>(workEnumerable);
+            return allWork;
+        }
+
+        public async Task<IReadOnlyDictionary<CatalogCommitItem, PackageDetailsCatalogLeaf>> GetEntryToLeafAsync(
```
Comment (loic-sharma) [inline src/NuGet.Services.V3/CommitCollectorUtility.cs]: Should we make it clear this doesn't load delete leaves? Maybe a name like `GetEntryToDetailsLeaf`? I don't feel strongly about this

File: src/NuGet.Services.V3/CommitCollectorUtility.cs [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+                                    work.PackageIdentity.Id,
+                                    work.PackageIdentity.Version.ToNormalizedString(),
+                                    work.Uri.AbsoluteUri);
+
+                                var leaf = await _catalogClient.GetPackageDetailsLeafAsync(work.Uri.AbsoluteUri);
+                                output.Add(KeyValuePair.Create(work, leaf));
+                            }
+                            catch (Exception ex)
+                            {
+                                _logger.LogError(
```
Comment (loic-sharma) [inline src/NuGet.Services.V3/CommitCollectorUtility.cs]: The latest dev branch has 2.2 logger. If you pull that in you should be able to drop the `0`.

File: src/NuGet.Services.AzureSearch/Catalog2AzureSearch/Catalog2AzureSearchCommand.cs (hunk 1 above)
Comment (joelverhagen) [reply src/NuGet.Services.AzureSearch/Catalog2AzureSearch/Catalog2AzureSearchCommand.cs]: Yes

File: src/NuGet.Services.V3/CommitCollectorUtility.cs (hunk 5 above)
Comment (joelverhagen) [reply src/NuGet.Services.V3/CommitCollectorUtility.cs]: Done.

File: src/NuGet.Services.V3/CommitCollectorUtility.cs (hunk 4 above)
Comment (joelverhagen) [reply src/NuGet.Services.V3/CommitCollectorUtility.cs]: Done.

File: src/NuGet.Services.V3/CommitCollectorUtility.cs (hunk 3 above)
Comment (joelverhagen) [reply src/NuGet.Services.V3/CommitCollectorUtility.cs]: Chose not to do this to minimize code changes while moving code.

**Pull request merged**